File: Python/第一章/07-model-boundary-demo/temperature_demo.py
"""对比低 temperature 与高 temperature 的生成结果。"""

import json
import os
import sys
from typing import Any, Callable
from urllib.error import HTTPError
from urllib.request import Request, urlopen


prompt = "请为退款审核助手写几句简短的欢迎语，只输出欢迎语。"
ChatRequester = Callable[[dict[str, Any], str], dict[str, Any]]


def request_chat_completion(
    request_body: dict[str, Any], api_key: str
) -> dict[str, Any]:
    """使用标准库调用 DeepSeek Chat Completions API。"""
    request = Request(
        "https://api.deepseek.com/chat/completions",
        data=json.dumps(request_body, ensure_ascii=False).encode("utf-8"),
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    try:
        with urlopen(request, timeout=60) as response:
            return json.loads(response.read().decode("utf-8"))
    except HTTPError as error:
        error_text = error.read().decode("utf-8", errors="replace")
        raise RuntimeError(
            f"DeepSeek API 调用失败：{error.code} {error_text}"
        ) from error


def run_group(
    temperature: float,
    *,
    api_key: str,
    requester: ChatRequester = request_chat_completion,
) -> list[str]:
    """使用指定 temperature 连续调用三次模型。"""
    print(f"\n===== temperature = {temperature} =====")
    outputs: list[str] = []

    for index in range(1, 4):
        result = requester(
            {
                "model": os.getenv("DEEPSEEK_MODEL", "deepseek-v4-flash"),
                "messages": [{"role": "user", "content": prompt}],
                "max_tokens": 80,
                "temperature": temperature,
                "stream": False,
                "thinking": {"type": "disabled"},
            },
            api_key,
        )
        try:
            content = result["choices"][0]["message"]["content"]
        except (KeyError, IndexError, TypeError) as error:
            raise RuntimeError("DeepSeek API 没有返回可用的回答。") from error

        print(f"{index}. {content}")
        outputs.append(content)

    return outputs


def run_demo(
    *,
    api_key: str | None = None,
    requester: ChatRequester = request_chat_completion,
) -> dict[float, list[str]]:
    """分别运行低温度与高温度实验。"""
    resolved_api_key = api_key or os.getenv("DEEPSEEK_API_KEY")
    if not resolved_api_key:
        raise RuntimeError("没有检测到 DEEPSEEK_API_KEY，请先在 .env 中配置。")

    return {
        0.1: run_group(0.1, api_key=resolved_api_key, requester=requester),
        1.5: run_group(1.5, api_key=resolved_api_key, requester=requester),
    }


def main() -> int:
    """运行 temperature 对比实验。"""
    try:
        run_demo()
    except (RuntimeError, OSError) as error:
        print(error, file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Node/第一章/07-model-boundary-demo/temperature-demo.js
const model = process.env.DEEPSEEK_MODEL ?? 'deepseek-v4-flash'

if (!process.env.DEEPSEEK_API_KEY) {
	console.error('没有检测到 DEEPSEEK_API_KEY，请先在 .env 中配置。')
	process.exit(1)
}

const prompt = '请为退款审核助手写几句简短的欢迎语，只输出欢迎语。'

/**
 * 使用指定的 temperature 连续调用三次模型。
 */
async function runGroup(temperature) {
	console.log(`\n===== temperature = ${temperature} =====`)

	for (let index = 1; index <= 3; index += 1) {
		const response = await fetch('https://api.deepseek.com/chat/completions', {
			method: 'POST',
			headers: {
				Authorization: `Bearer ${process.env.DEEPSEEK_API_KEY}`,
				'Content-Type': 'application/json'
			},
			body: JSON.stringify({
				model,
				messages: [
					{
						role: 'user',
						content: prompt
					}
				],
				max_tokens: 80,
				temperature,
				stream: false,

				// DeepSeek V4 默认开启思考模式。
				// 官方文档明确说明：思考模式下 temperature 和 top_p 不会生效。
				// 所以这次研究 temperature 时，需要显式关闭思考模式。
				thinking: {
					type: 'disabled'
				}
			})
		})

		const result = await response.json()

		if (!response.ok) {
			console.error('DeepSeek API 返回错误：')
			console.dir(result, { depth: null })
			process.exit(1)
		}

		console.log(`${index}. ${result.choices[0].message.content}`)
	}
}

async function main() {
	// 低 temperature：生成的内容会更加集中，几次回答可能比较相似。
	await runGroup(0.1)
	// 高 temperature：模型更容易选择原本概率较低的 Token，所以回答通常会出现更多变化。
	await runGroup(1.5)
}

main().catch((error) => {
	console.error('实验执行失败：', error)
	process.exit(1)
})
